File: web/src/format.ts
const months = [
  "Jan",
  "Feb",
  "Mar",
  "Apr",
  "May",
  "Jun",
  "Jul",
  "Aug",
  "Sep",
  "Oct",
  "Nov",
  "Dec",
];

const two = (value: number) => String(value).padStart(2, "0");

export const clockTime = (date: Date) =>
  `${two(date.getHours())}:${two(date.getMinutes())}`;

export const dayMonth = (date: Date) =>
  `${two(date.getDate())} ${months[date.getMonth()]}`;

// Desktop mail-row date: the time today, otherwise day and month.
export function rowDate(date: Date, now = new Date()) {
  const sameDay =
    date.getFullYear() === now.getFullYear() &&
    date.getMonth() === now.getMonth() &&
    date.getDate() === now.getDate();
  return sameDay ? clockTime(date) : dayMonth(date);
}

// Desktop reader date line, shown above the time.
export const readerDate = (date: Date) =>
  `${dayMonth(date)} ${date.getFullYear()}`;

File: flutter/lib/ui/format.dart
const _months = [
  'Jan',
  'Feb',
  'Mar',
  'Apr',
  'May',
  'Jun',
  'Jul',
  'Aug',
  'Sep',
  'Oct',
  'Nov',
  'Dec',
];

String _two(int value) => value.toString().padLeft(2, '0');

String clockTime(DateTime date) => '${_two(date.hour)}:${_two(date.minute)}';

String dayMonth(DateTime date) =>
    '${_two(date.day)} ${_months[date.month - 1]}';

/// Desktop mail-row date: the time today, otherwise day and month.
String rowDate(DateTime date, {DateTime? now}) {
  final local = date.toLocal();
  final today = (now ?? DateTime.now()).toLocal();
  final sameDay =
      local.year == today.year &&
      local.month == today.month &&
      local.day == today.day;
  return sameDay ? clockTime(local) : dayMonth(local);
}

/// Desktop reader date: "09 Sep 2026" over the time.
String readerDate(DateTime date) {
  final local = date.toLocal();
  return '${dayMonth(local)} ${local.year}\n${clockTime(local)}';
}
